apm-agent-plugins/apm-scala-concurrent-plugin/src/main/java/co/elastic/apm/agent/scala/concurrent/FutureInstrumentation.java

```diff
@@ -0,0 +1,78 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.scala.concurrent;
+
+import co.elastic.apm.agent.bci.ElasticApmInstrumentation;
+import co.elastic.apm.agent.bci.VisibleForAdvice;
+import co.elastic.apm.agent.impl.transaction.TraceContextHolder;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nonnull;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static net.bytebuddy.matcher.ElementMatchers.*;
+
+public class FutureInstrumentation extends ElasticApmInstrumentation {
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return hasSuperType(named("scala.concurrent.Future"))
+            .or(hasSuperType(named("scala.concurrent.impl.Promise")))
+            .or(hasSuperType(named("scala.concurrent.impl.Promise$Transformation")))
+            .or(hasSuperType(named("scala.concurrent.Future$")));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("onComplete").and(returns(void.class))
+            .or(named("transform").and(returns(named("scala.concurrent.Future"))))
+            .or(named("transformWith").and(returns(named("scala.concurrent.Future"))))
+            .or(named("result"));
+    }
+
+    @Nonnull
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Arrays.asList("concurrent", "future");
+    }
+
+    @VisibleForAdvice
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void onEnter() {
+        System.out.println("DEBUG2");
```

Comment:
Im currently trying out to create a first real test to see how everything works.
Im currently stuck on why ByteBuddy is not executing this `onEnter` function even though I see method-matches on DEBUG level:
```
[ScalaTest-main-running-FutureInstrumentationSpec] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - Method match for instrumentation FutureInstrumentation: ((((name(equals(onComplete)) and returns(erasure(is(void)))) or (name(equals(transform)) and returns(erasure(name(equals(scala.concurrent.Future)))))) or (name(equals(transformWith)) and returns(erasure(name(equals(scala.concurrent.Future)))))) or name(equals(result))) matches public final void scala.concurrent.impl.Promise$DefaultPromise.onComplete(scala.Function1,scala.concurrent.ExecutionContext)
[ScalaTest-main-running-FutureInstrumentationSpec] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - Method match for instrumentation FutureInstrumentation: ((((name(equals(onComplete)) and returns(erasure(is(void)))) or (name(equals(transform)) and returns(erasure(name(equals(scala.concurrent.Future)))))) or (name(equals(transformWith)) and returns(erasure(name(equals(scala.concurrent.Future)))))) or name(equals(result))) matches public final java.lang.Object scala.concurrent.impl.Promise$DefaultPromise.result(scala.concurrent.duration.Duration,scala.concurrent.CanAwait) throws java.lang.Exception
```

I see that the log also mentions `throws java.lang.Exception` but it is unclear on how and when that Exception is thrown.
@felixbarny Any idea on what I'm missing on why `onEnter` is not being called even though there is a method match?

Reply:
Have you tried setting a breakpoint at `Promise$DefaultPromise.onComplete` or a method breakpoint for `Promise.onComplete` to see if the methods get called at all?

> I see that the log also mentions throws java.lang.Exception but it is unclear on how and when that Exception is thrown.

This is just a part of the method signature to indicate this method may throw any `java.lang.Exception`.

Reply:
Thanks for the tip, this helped me figure out the underlying process a lot better :)

Struggling with ByteBuddy now to figure out how I could add a transaction-id to a matched class as a field.
From my current understanding, a `Promise` will be created and at some point, the `run` method will be called on that `Promise`. Since the call on that `run` could happen on a different thread, I think it would make sense to somehow store the original context on that `Promise` and make sure to activate it again before the `run`.

I've tried working with `@Advice.FieldValue` and `@Advice.Local` but I'm not sure if those values are really kept on the `Promise` itself.

Reply:
Our agent supports to be attached at runtime. One of the limitations when instrumenting classes that are already loaded is that you can't change the class file format. That means you can't add interfaces, methods or fields.

A common technique we use to overcome this is to put things like a `Promise` as the key of a `WeakConcurrentMap` and the corresponding `Span` as the value.

Example:
https://github.com/elastic/apm-agent-java/blob/842185664456651f1d6a99602d21a7c052ddf111/apm-agent-plugins/apm-redis-plugin/apm-lettuce-plugin/src/main/java/co/elastic/apm/agent/redis/lettuce/Lettuce5StartSpanInstrumentation.java#L57

Reply:
What if it is a Scala `Promise`? I remember a comment that we should prevent loading Scala classes since they would not be available on all code-bases.
Especially, I'm not sure if Java has any Scala interop.

Reply:
Defining a `WeakConcurrentMap<Promise, Span>` does not cause loading/linking the `Promise`.  That's because the javac compiler erases the generic arguments. We only put promises in at runtime as we encounter these promises so they are guaranteed to already have been loaded.